source/common/upstream/health_checker_impl.cc

```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
   request_.asArray().swap(values);
 }
 
+GrpcHealthCheckerImpl::GrpcHealthCheckerImpl(const Cluster& cluster,
+                                             const envoy::api::v2::HealthCheck& config,
+                                             Event::Dispatcher& dispatcher,
+                                             Runtime::Loader& runtime,
+                                             Runtime::RandomGenerator& random)
+    : HealthCheckerImplBase(cluster, config, dispatcher, runtime, random),
+      service_method_(*Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+          "grpc.health.v1.Health.Check")) {
+  if (!config.grpc_health_check().service_name().empty()) {
+    service_name_.value(config.grpc_health_check().service_name());
+  }
+}
+
+GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::GrpcActiveHealthCheckSession(
+    GrpcHealthCheckerImpl& parent, HostSharedPtr host)
+    : ActiveHealthCheckSession(parent, host), parent_(parent) {}
+
+GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::~GrpcActiveHealthCheckSession() {
+  if (client_) {
+    // If there is an active request it will get reset, so make sure we ignore the reset.
+    expect_reset_ = true;
+    client_->close();
+  }
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::decodeHeaders(
+    Http::HeaderMapPtr&& headers, bool end_stream) {
+  const auto http_response_status = Http::Utility::getResponseStatus(*headers);
+  if (http_response_status != enumToInt(Http::Code::OK)) {
+    // https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md requires that
+    // grpc-status be used if available.
+    if (end_stream) {
+      auto grpc_status = Grpc::Common::getGrpcStatus(*headers);
+      if (grpc_status.valid()) {
+        onRpcComplete(grpc_status.value(), Grpc::Common::getGrpcMessage(*headers), true);
+        return;
+      }
+    }
+    onRpcComplete(Grpc::Common::httpToGrpcStatus(http_response_status), "non-200 HTTP response",
+                  end_stream);
+    return;
+  }
+  if (!Grpc::Common::hasGrpcContentType(*headers)) {
+    onRpcComplete(Grpc::Status::GrpcStatus::Internal, "invalid gRPC content-type", false);
+    return;
+  }
+  if (end_stream) {
+    // This is how, for instance, grpc-go signals about missing service - HTTP/2 200 OK with
+    // 'unimplemented' gRPC status.
+    auto grpc_status = Grpc::Common::getGrpcStatus(*headers);
+    if (grpc_status.valid()) {
+      onRpcComplete(grpc_status.value(), Grpc::Common::getGrpcMessage(*headers), true);
+      return;
+    }
+    onRpcComplete(Grpc::Status::GrpcStatus::Internal, "unexpected stream end", true);
```

Comment:
This is a gRPC protocol violation, so maybe be a bit more explicit about that.

Reply:
Not sure what to do here then... there is no more suitable grpc status. I can add "protocol violation: " to error message.

Reply:
Yeah, just add to the error message.